### ISSUE TITLE
The @ syntax has issues with Component slices

### ISSUE THREAD

**Issue by rokf** [a-h/templ#414]
Hello,

I use `templ.Component` slices (`[]templ.Component`) in some of my `templ` files. If I try to use a loop to go over them and use the index (integer) to access the components in the slice the `@` syntax breaks.

`@items[i]` doesn't work.

`{! items[i] }` works.

My `templ` version is `v0.2.513`. My `go` version is `go1.21.6 darwin/arm64`.

The `templ` file formatter in VSCode also does a weird thing and puts the `[i]` part on a line below the `@items` part, which results in the following two compilation errors:

```
i declared and not used
items.Render undefined (type []templ.Component has no field or method Render)
```

I can rewrite the code with the `@` syntax in such a way that it'll work again, that isn't a problem. **I'm wondering if this is perhaps a bug or if `@` isn't meant to be used with the slice indexes?**

Thanks!

**Comment by joerdav** [comment]
I agree that this should be fixed, it could either be classified as a bug or undetermined behavior!

This issue is that currently the supported templ element expression parser (the `@` syntax) doesn't actually parse as a go expression, it is a separate parser that checks for a string that looks a bit like `@template` or `@templates.MyTemplate` or `@template(params)`.

See the parser here: https://github.com/a-h/templ/blob/main/parser/v2/templelementparser.go#L17-L25 

The 2 fixes for this would be (in order of increasing complexity): 
- Expand the existing parser to allow indexing `[` + goExpression + `]`
- Rewrite the existing parser to parse any Go expression after the `@`

**Comment by benjiro29** [comment]
Ran into the same issue ...

Imaging running templ fmt on all your code "as advised" and seeing it all break because of this issue. Last time it took me several hours to reverse the changes and going back to a older version.

There really seems to be a lack of proper testing before changes are pushed to the public.  I do not understand why the push for @ when the old {! } syntax worked perfectly (and is more readable, imho).

**Comment by joerdav** [comment]
I also took a look at this. I initially went for a more general parser that parsed from `@` to the end of the line, or occurrence of `{\n`. This passed the tests mostly, but upon a search of templ files across github it would have broken occurrences where templelement params are split across multiple lines (I plan on adding tests for this to lock in the behavior).

So, I defaulted to the dumb solution of just allowing an index pattern in a call expression, but your solution @a-h is definitely the proper way to do this. With mine it would have been a matter of time until someone had another way of rendering an element that they wanted supporting.

**Comment by a-h** [comment]
I've done some a few more updates to https://github.com/a-h/expression/ to add support for `if`, `for` and `else` statements. The results are promising in that the parser would be more robust if I replaced the current expression code with this method (aside from the JavaScript expression parser).

**Comment by a-h** [comment]
@benjiro29 - if you want to help, when I've got the expression parser more complete, additional test cases for the expression parsers would be good.

**Comment by a-h** [comment]
I'm very comfortable with the Go expression parser now. Next I'm going to copy the https://github.com/a-h/expression code into a package within the templ parser, and migrate all of templ's Go expression parsing to using it instead.

**Comment by a-h** [comment]
I've got a wip branch here: https://github.com/a-h/templ/tree/go_ast

It lays out the changes that need to be made. It's not quite working yet - I need to make a few more changes to resolve the broken tests (the issue is the implementation, not the tests 😁 ), and add tests for the function declaration AST parser in the goparser directory.

**Comment by a-h** [comment]
The WIP branch is now https://github.com/a-h/templ/pull/461

Reviews welcome, additional test cases that test valid edge cases that I may not have considered also welcome.

**Comment by joerdav** [comment]
Fixed in #461
